### ISSUE TITLE
Discuss allowing extensions to reliably draw over pages

### ISSUE THREAD

**Issue by oliverdunk** [hober/web-extensions#235]
Many extensions have a need to draw UI over the page. Developers have traditionally relied on the [stacking context](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Positioning/Understanding_z_index/The_stacking_context) and specific properties like `z-index`. These have always been limited but for the most part, do the job.

With the introduction of modals (via the `<dialog>` element), web developers can now draw to a new "top layer" that renders above everything else. Example: https://codepen.io/oliverdunk/pen/xxWKjOZ. This makes it much harder for extension developers as they now have to ensure UI is inside of the highest modal, and this could involve repeatedly moving their UI between DOM elements.

The ideal solution here would be a standardized mechanism - maybe a new layer - that extensions could put UI on. This would be reliable even as new standards are developed and provide the most robust long-term solution.

**Comment by oliverdunk** [comment]
From @dotproto:

> I just poked through the HTML spec a bit to read up on the Dialog element. [showModal()](https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-showmodal) touches on a couple of important concepts for the sake of this discussion:
> 
> - Step 5 sets the [is modal](https://html.spec.whatwg.org/multipage/interactive-elements.html#is-modal) flag of the dialog to true
> - Step 6 [blocks](https://html.spec.whatwg.org/multipage/interaction.html#blocked-by-a-modal-dialog) the dialog's node document
> - Step 7 adds the dialog to the node document's [top layer](https://fullscreen.spec.whatwg.org/#top-layer)

**Comment by jonathanKingston** [comment]
> I just poked through the HTML spec a bit to read up on the Dialog element. [showModal()](https://html.spec.whatwg.org/multipage/interactive-elements.html#dom-dialog-showmodal) touches on a couple of important concepts for the sake of this discussion

Dialog showModal has a lot of the properties that extensions want; however the page can call showModal after the extension and essentially click-jack the interface the extension wishes to present to the user.
- Ideally the extension could prevent further modals whilst theirs is open.

**Comment by birtles** [comment]
> This makes it much harder for extension developers as they now have to ensure UI is inside of the highest modal, and this could involve repeatedly moving their UI between DOM elements.

Note that this is already the case for the full screen API -- that is, an extension needs to listen for fullscreen events and move their UI to be a descendant of `document.fullscreenElement` and back.

**Comment by Rob--W** [comment]
The concept of overlaying content on top of the current page (not necessarily embedded in the page) has come up many times before. Here's a feature request in Firefox for the generic functionality without any prescribed mechanism: https://bugzilla.mozilla.org/show_bug.cgi?id=1340930

Question to the extension devs here: Would `action.openPopup` cover the use cases (described in more detail at #160)? The extension popup panel can then host the content. From the user's perspective, the panel is clearly attributed to a specific extension.

**Comment by oliverdunk** [comment]
I think there are distinct use cases which aren't solved by openPopup, and would benefit greatly from the sort of thing proposed in that issue. In the case of 1Password, we want to show UI directly below input fields as this is an established UI pattern for offering field suggestions. We also have cases where we want to show notifications offering the user a suggested action without stealing focus away from the page.

It feels like drawing to the screen is a big part of allowing extensions to innovate. I could imagine, for example, an accessibility extension which highlights regions of the page to help someone navigate.

Something at a higher level than the main DOM would be really exciting because as @jonathanKingston mentioned, there's real scope for clickjacking issues when extensions are forced to put themselves in the DOM and compete with other scripts on the page to make sure they appear highest in the stacking order.

**Comment by birtles** [comment]
> In the case of 1Password, we want to show UI directly below input fields as this is an established UI pattern for offering field suggestions. We also have cases where we want to show notifications offering the user a suggested action without stealing focus away from the page.

Pop-up dictionaries have similar requirements. That is, they need to pop-up next to or below the word being selected/hovered and typically without affecting the page's focus.

**Comment by hanguokai** [comment]
There is a new experiment(origin trial) Web API: [Pop-up](https://developer.chrome.com/blog/pop-ups-theyre-making-a-resurgence/). Maybe this API can solve the problem in web pages.

Outside of web pages, like `action.openPopup`, new [`chrome.sidePanel`](https://bugs.chromium.org/p/chromium/issues/detail?id=1328645) api may also solve the problem for some cases.

**Comment by oliverdunk** [comment]
I opened an issue about this in the open-ui group (https://github.com/openui/open-ui/issues/626) since I was curious to see if they wanted to tackle it together. They agreed it was an interesting problem but don't have the bandwidth to tackle it, so I think this is a good one to continue working on ourselves.

IIRC the action item from the last time this was discussed at a meeting was putting together an API proposal.

**Comment by fregante** [comment]
As discussed in today's meeting, browsers are already able to draw over pages via the dev tools. It would be great to expose this capability without having to alter the host page.

Example: Safari's dev tools drawn on this very page.

<img width="290" alt="Screenshot 7" src="https://github.com/user-attachments/assets/32a9413e-03fa-4336-a876-251566e9f89a">

If this UI is painted via a standard "transparent" `document`, it might be good to expose it either via `browser.dom.getOverlayDocument()` or, as Firefox does in a [somewhat related scenario](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts), via `window.overlayDocument`. From there, the standard DOM API applies (with CSP to exclude scripting)

**Comment by jonathanKingston** [comment]
https://www.techspot.com/news/109149-lastpass-1password-bitwarden-extensions-vulnerable-clickjacking-attacks.html outlines attacks against all common password managers.

I think there's value in prioritising the standardisation here.

**Comment by craiglurey** [comment]
Hi @oliverdunk and friends, let's set up a call and review some solutions.

**Comment by frivain-dashlane** [comment]
Agreed. I'd like for us to reopen that conversation and look for options.

**Comment by ylemkimon** [comment]
This seems to have been discussed in [Click-jacking Protection](https://github.com/w3c/webextensions/blob/main/_minutes/2026-04-07-london-f2f.md#click-jacking-protection) (#979).

I think there needs to be a way for extension to provide a secure UI surface, that the page renderer process cannot access.
